pkg/tfgen/convert_cli.go

```diff
@@ -0,0 +1,427 @@
+	contract.AssertNoErrorf(err, "pulumiParser.ParseFile returned an error")
+
+	var diagnostics hcl.Diagnostics
+
+	diagnostics = append(diagnostics, pulumiParser.Diagnostics...)
+	if diagnostics.HasErrors() {
+		return "", diagnostics
+	}
+
+	opts := cc.opts
+	if opts == nil {
+		var opts []pcl.BindOption
+		opts = append(opts, pcl.AllowMissingProperties)
+		opts = append(opts, pcl.AllowMissingVariables)
+		opts = append(opts, pcl.SkipResourceTypechecking)
+		if cc.pluginHost != nil {
+			opts = append(opts, pcl.PluginHost(cc.pluginHost))
+			loader := newLoader(cc.pluginHost)
+			loader.BindCurrentPackage(*spec, cc.packageName)
+			opts = append(opts, pcl.Loader(loader))
+		}
+		if cc.packageCache != nil {
+			opts = append(opts, pcl.Cache(cc.packageCache))
+		}
+		cc.opts = opts
+	}
+
+	program, programDiags, err := pcl.BindProgram(pulumiParser.Files, opts...)
+	if err != nil {
+		diagnostics = append(diagnostics, &hcl.Diagnostic{
+			Severity: hcl.DiagError,
+			Summary:  "pcl.BindProgram failed",
+			Detail:   source,
+		})
+		return "", diagnostics
+	}
+
+	diagnostics = append(diagnostics, programDiags...)
+	if diagnostics.HasErrors() {
+		return "", diagnostics
+	}
+
+	var genDiags hcl.Diagnostics
+	var generatedFiles map[string][]byte
+
+	switch languageName {
+	case "typescript":
+		generatedFiles, genDiags, err = hcl2nodejs.GenerateProgram(program)
+		diagnostics = append(diagnostics, genDiags...)
+	case "python":
+		generatedFiles, genDiags, err = hcl2python.GenerateProgram(program)
+		diagnostics = append(diagnostics, genDiags...)
+	case "csharp":
+		generatedFiles, genDiags, err = hcl2dotnet.GenerateProgram(program)
+		diagnostics = append(diagnostics, genDiags...)
+	case "go":
+		generatedFiles, genDiags, err = hcl2go.GenerateProgram(program)
+		diagnostics = append(diagnostics, genDiags...)
+	case "yaml":
+		generatedFiles, genDiags, err = hcl2yaml.GenerateProgram(program)
+		diagnostics = append(diagnostics, genDiags...)
+	case "java":
+		generatedFiles, genDiags, err = hcl2java.GenerateProgram(program)
+		diagnostics = append(diagnostics, genDiags...)
+	default:
+		err = fmt.Errorf("Unsupported language: %q", languageName)
+	}
+
+	contract.AssertNoErrorf(err, "language specific generator returned an error")
+	if len(generatedFiles) != 1 {
+		diagnostics = append(diagnostics, &hcl.Diagnostic{
+			Severity: hcl.DiagError,
+			Summary:  "GenerateProgram failed",
+			Detail: fmt.Sprintf("expected 1 file to be generated, got %d",
+				len(generatedFiles)),
+		})
+		return "", diagnostics
+	}
+	var key string
+	for n := range generatedFiles {
+		key = n
+	}
+	return string(generatedFiles[key]), diagnostics
+}
+
+// Encodes an example conversion request into a special form that needs to survive being encoded
+// into JSON and plain text and recovered back later on.
+func (cc *cliConverter) encodeConversionJob(hcl string, languageName string) string {
+	return fmt.Sprintf("{example:%s:%s}", languageName,
```

Comment:
This is a creative solution that avoids refactoring too much code to achieve a batched/bulk interface with pulumi convert. The examples instead of getting rendered get spliced as is. This splice travels all the way into the generated PackageSpec JSON. A subsequent pass regexes it out and actually performs bulk conversion.